File: swift-bridge/Sources/CoreAnimationBridge/CAMediaTiming.swift
import Foundation
import QuartzCore

func caTimingFunctionName(_ raw: Int32) -> CAMediaTimingFunctionName {
    switch raw {
    case 0: return .linear
    case 1: return .easeIn
    case 2: return .easeOut
    case 3: return .easeInEaseOut
    default: return .default
    }
}

private func caTimingFunctionMatches(
    _ lhs: CAMediaTimingFunction,
    _ rhs: CAMediaTimingFunction
) -> Bool {
    for index in 0..<4 {
        var left = [Float](repeating: 0, count: 2)
        var right = [Float](repeating: 0, count: 2)
        lhs.getControlPoint(at: index, values: &left)
        rhs.getControlPoint(at: index, values: &right)
        if abs(left[0] - right[0]) > 0.000_1 || abs(left[1] - right[1]) > 0.000_1 {
            return false
        }
    }
    return true
}

func caTimingFunctionNameRaw(_ function: CAMediaTimingFunction?) -> Int32 {
    guard let function else { return -1 }
    let mappings: [(Int32, CAMediaTimingFunctionName)] = [
        (0, .linear),
        (1, .easeIn),
        (2, .easeOut),
        (3, .easeInEaseOut),
        (4, .default),
    ]
    for (raw, name) in mappings {
        if caTimingFunctionMatches(function, CAMediaTimingFunction(name: name)) {
            return raw
        }
    }
    return -1
}

@_cdecl("ca_timing_function_new_named")
public func ca_timing_function_new_named(_ value: Int32) -> UnsafeMutableRawPointer? {
    caRetain(CAMediaTimingFunction(name: caTimingFunctionName(value)))
}

@_cdecl("ca_timing_function_new_control_points")
public func ca_timing_function_new_control_points(_ c1x: Float, _ c1y: Float, _ c2x: Float, _ c2y: Float) -> UnsafeMutableRawPointer? {
    caRetain(CAMediaTimingFunction(controlPoints: c1x, c1y, c2x, c2y))
}

@_cdecl("ca_timing_function_get_name")
public func ca_timing_function_get_name(_ handle: UnsafeMutableRawPointer?) -> Int32 {
    guard let function: CAMediaTimingFunction = caBorrow(handle) else { return -1 }
    return caTimingFunctionNameRaw(function)
}

@_cdecl("ca_timing_function_get_control_point")
public func ca_timing_function_get_control_point(_ handle: UnsafeMutableRawPointer?, _ index: Int, _ outValues: UnsafeMutableRawPointer?) -> Bool {
    guard let function: CAMediaTimingFunction = caBorrow(handle), let outValues, index >= 0, index < 4 else {
        return false
    }
    var values = [Float](repeating: 0, count: 2)
    function.getControlPoint(at: index, values: &values)
    let ptr = outValues.assumingMemoryBound(to: Float.self)
    ptr[0] = values[0]
    ptr[1] = values[1]
    return true
}

func caFillMode(_ raw: Int32) -> CAMediaTimingFillMode {
    switch raw {
    case 1: return .forwards
    case 2: return .backwards
    case 3: return .both
    default: return .removed
    }
}

func caFillModeRaw(_ value: CAMediaTimingFillMode) -> Int32 {
    switch value {
    case .forwards: return 1
    case .backwards: return 2
    case .both: return 3
    default: return 0
    }
}

@_cdecl("ca_animation_get_begin_time")
public func ca_animation_get_begin_time(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let animation: CAAnimation = caBorrow(handle) else { return 0 }
    return animation.beginTime
}

@_cdecl("ca_animation_set_begin_time")
public func ca_animation_set_begin_time(_ handle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let animation: CAAnimation = caBorrow(handle) else { return }
    animation.beginTime = value
}

@_cdecl("ca_animation_get_speed")
public func ca_animation_get_speed(_ handle: UnsafeMutableRawPointer?) -> Float {
    guard let animation: CAAnimation = caBorrow(handle) else { return 0 }
    return animation.speed
}

@_cdecl("ca_animation_set_speed")
public func ca_animation_set_speed(_ handle: UnsafeMutableRawPointer?, _ value: Float) {
    guard let animation: CAAnimation = caBorrow(handle) else { return }
    animation.speed = value
}

@_cdecl("ca_animation_get_time_offset")
public func ca_animation_get_time_offset(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let animation: CAAnimation = caBorrow(handle) else { return 0 }
    return animation.timeOffset
}

@_cdecl("ca_animation_set_time_offset")
public func ca_animation_set_time_offset(_ handle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let animation: CAAnimation = caBorrow(handle) else { return }
    animation.timeOffset = value
}

@_cdecl("ca_animation_get_repeat_duration")
public func ca_animation_get_repeat_duration(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let animation: CAAnimation = caBorrow(handle) else { return 0 }
    return animation.repeatDuration
}

@_cdecl("ca_animation_set_repeat_duration")
public func ca_animation_set_repeat_duration(_ handle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let animation: CAAnimation = caBorrow(handle) else { return }
    animation.repeatDuration = value
}

@_cdecl("ca_animation_get_fill_mode")
public func ca_animation_get_fill_mode(_ handle: UnsafeMutableRawPointer?) -> Int32 {
    guard let animation: CAAnimation = caBorrow(handle) else { return 0 }
    return caFillModeRaw(animation.fillMode)
}

@_cdecl("ca_animation_set_fill_mode")
public func ca_animation_set_fill_mode(_ handle: UnsafeMutableRawPointer?, _ value: Int32) {
    guard let animation: CAAnimation = caBorrow(handle) else { return }
    animation.fillMode = caFillMode(value)
}

File: src/ca_media_timing.rs
use crate::animation::Animation;
use crate::private::handle_type;
use crate::transaction::Transaction;

handle_type!(TimingFunction);

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(i32)]
pub enum TimingFunctionName {
    Linear = 0,
    EaseIn = 1,
    EaseOut = 2,
    EaseInEaseOut = 3,
    Default = 4,
}

impl TimingFunctionName {
    pub(crate) const fn from_raw(value: i32) -> Option<Self> {
        match value {
            0 => Some(Self::Linear),
            1 => Some(Self::EaseIn),
            2 => Some(Self::EaseOut),
            3 => Some(Self::EaseInEaseOut),
            4 => Some(Self::Default),
            _ => None,
        }
    }

    pub(crate) const fn raw(self) -> i32 {
        self as i32
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(i32)]
pub enum MediaTimingFillMode {
    Removed = 0,
    Forwards = 1,
    Backwards = 2,
    Both = 3,
}

impl MediaTimingFillMode {
    pub(crate) const fn from_raw(value: i32) -> Self {
        match value {
            1 => Self::Forwards,
            2 => Self::Backwards,
            3 => Self::Both,
            _ => Self::Removed,
        }
    }
}

impl TimingFunction {
    #[must_use]
    pub fn with_name(name: TimingFunctionName) -> Option<Self> {
        unsafe { Self::from_raw(crate::ffi::ca_timing_function_new_named(name.raw())) }
    }

    #[must_use]
    pub fn with_control_points(c1x: f32, c1y: f32, c2x: f32, c2y: f32) -> Option<Self> {
        unsafe {
            Self::from_raw(crate::ffi::ca_timing_function_new_control_points(
                c1x, c1y, c2x, c2y,
            ))
        }
    }

    #[must_use]
    pub fn name(&self) -> Option<TimingFunctionName> {
        TimingFunctionName::from_raw(unsafe {
            crate::ffi::ca_timing_function_get_name(self.as_ptr())
        })
    }

    #[must_use]
    pub fn control_point(&self, index: usize) -> Option<(f32, f32)> {
        let mut values = [0.0_f32; 2];
        let ok = unsafe {
            crate::ffi::ca_timing_function_get_control_point(
                self.as_ptr(),
                index,
                values.as_mut_ptr().cast::<core::ffi::c_void>(),
            )
        };
        ok.then_some((values[0], values[1]))
    }
}

impl Animation {
    #[must_use]
    pub fn timing_function(&self) -> Option<TimingFunction> {
        unsafe {
            TimingFunction::from_raw(crate::ffi::ca_animation_get_timing_function(self.as_ptr()))
        }
    }

    pub fn set_timing_function(&self, value: Option<&TimingFunction>) {
        unsafe {
            crate::ffi::ca_animation_set_timing_function(
                self.as_ptr(),
                value.map_or(core::ptr::null_mut(), TimingFunction::as_ptr),
            )
        };
    }

    #[must_use]
    pub fn begin_time(&self) -> f64 {
        unsafe { crate::ffi::ca_animation_get_begin_time(self.as_ptr()) }
    }

    pub fn set_begin_time(&self, value: f64) {
        unsafe { crate::ffi::ca_animation_set_begin_time(self.as_ptr(), value) };
    }

    #[must_use]
    pub fn speed(&self) -> f32 {
        unsafe { crate::ffi::ca_animation_get_speed(self.as_ptr()) }
    }

    pub fn set_speed(&self, value: f32) {
        unsafe { crate::ffi::ca_animation_set_speed(self.as_ptr(), value) };
    }

    #[must_use]
    pub fn time_offset(&self) -> f64 {
        unsafe { crate::ffi::ca_animation_get_time_offset(self.as_ptr()) }
    }

    pub fn set_time_offset(&self, value: f64) {
        unsafe { crate::ffi::ca_animation_set_time_offset(self.as_ptr(), value) };
    }

    #[must_use]
    pub fn repeat_duration(&self) -> f64 {
        unsafe { crate::ffi::ca_animation_get_repeat_duration(self.as_ptr()) }
    }

    pub fn set_repeat_duration(&self, value: f64) {
        unsafe { crate::ffi::ca_animation_set_repeat_duration(self.as_ptr(), value) };
    }

    #[must_use]
    pub fn fill_mode(&self) -> MediaTimingFillMode {
        MediaTimingFillMode::from_raw(unsafe {
            crate::ffi::ca_animation_get_fill_mode(self.as_ptr())
        })
    }

    pub fn set_fill_mode(&self, value: MediaTimingFillMode) {
        unsafe { crate::ffi::ca_animation_set_fill_mode(self.as_ptr(), value as i32) };
    }
}

impl Transaction {
    #[must_use]
    pub fn animation_timing_function() -> Option<TimingFunction> {
        unsafe {
            TimingFunction::from_raw(crate::ffi::ca_transaction_get_animation_timing_function())
        }
    }

    pub fn set_animation_timing_function(value: Option<&TimingFunction>) {
        unsafe {
            crate::ffi::ca_transaction_set_animation_timing_function(
                value.map_or(core::ptr::null_mut(), TimingFunction::as_ptr),
            )
        };
    }
}
